idmtest/idmtest.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package idmtest holds a mock implementation of the identity manager
+// suitable for testing.
+package idmtest
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"sync"
+
+	"github.com/juju/httprequest"
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery/agent"
+
+	"github.com/juju/identity/params"
+)
+
+// Server represents a mock identity server.
+// It currently serves only the discharge and groups endpoints.
+type Server struct {
+	// URL holds the URL of the mock identity server.
+	// The discharger endpoint is located at URL/v1/discharge.
+	URL *url.URL
+
+	// PublicKey holds the public key of the mock identity server.
+	PublicKey *bakery.PublicKey
+
+	router *httprouter.Router
+	srv    *httptest.Server
+	bakery *bakery.Service
+
+	// mu guards the fields below it.
+	mu          sync.Mutex
+	users       map[string]*user
+	defaultUser string
+	waits       []chan struct{}
+}
+
+type user struct {
+	groups []string
+	key    *bakery.KeyPair
+}
+
+// NewServer runs a mock identity server. It can discharge
+// macaroons and return information on user group membership.
+// The returned server should be closed after use.
+func NewServer() *Server {
+	srv := &Server{
+		users: make(map[string]*user),
+	}
+	bsvc, err := bakery.NewService(bakery.NewServiceParams{
+		Locator: srv,
+	})
+	if err != nil {
+		panic(err)
+	}
+	srv.bakery = bsvc
+	srv.PublicKey = bsvc.PublicKey()
+	errorMapper := httprequest.ErrorMapper(errToResp)
+	h := &handler{
+		srv: srv,
+	}
+	router := httprouter.New()
+	for _, route := range errorMapper.Handlers(func(httprequest.Params) (*handler, error) {
+		return h, nil
+	}) {
+		router.Handle(route.Method, route.Path, route.Handle)
+	}
+	mux := http.NewServeMux()
+	httpbakery.AddDischargeHandler(mux, "/v1/discharger", srv.bakery, srv.check)
+	router.Handler("POST", "/v1/discharger/*rest", mux)
```

Comment:
There is an extension for this to allow for the /publickey and /discharge endpoints. We need to make sure it ends up here (not in this PR though).